tests/python/pants_test/backend/codegen/tasks/test_simple_codegen_task.py

```diff
@@ -0,0 +1,230 @@
+# coding=utf-8
+# Copyright 2015 Pants project contributors (see CONTRIBUTORS.md).
+# Licensed under the Apache License, Version 2.0 (see LICENSE).
+
+from __future__ import (absolute_import, division, generators, nested_scopes, print_function,
+                        unicode_literals, with_statement)
+
+import os
+from textwrap import dedent
+
+from pants.backend.codegen.register import build_file_aliases as register_codegen
+from pants.backend.codegen.tasks.simple_codegen_task import SimpleCodegenTask
+from pants.backend.core.register import build_file_aliases as register_core
+from pants.backend.jvm.targets.java_library import JavaLibrary
+from pants.backend.jvm.targets.jvm_target import JvmTarget
+from pants.base.build_file_aliases import BuildFileAliases
+from pants.base.payload import Payload
+from pants_test.tasks.task_test_base import TaskTestBase
+
+
+class SimpleCodegenTaskTest(TaskTestBase):
+  @classmethod
+  def task_type(cls):
+    return SimpleCodegenTask
+
+  @property
+  def alias_groups(self):
+      return register_core().merge(register_codegen()).merge(BuildFileAliases.create({
+        'dummy_library': SimpleCodegenTaskTest.DummyLibrary
+      }))
+
+  def _create_subclass_task(self, subclass_type, target_roots=None, **options):
+    scope = 'test_scope_{}'.format(type(subclass_type))
+    task_type = self.synthesize_task_subtype(subclass_type, scope)
+    self.set_options_for_scope(scope, **options)
+    context = self.context(for_task_types=[task_type], target_roots=target_roots)
+    return task_type(context, self._test_workdir)
+
+  def _create_dummy_task(self, target_roots, **options):
+    SimpleCodegenTaskTest.DummyGen.TEST_BASE = self
+    SimpleCodegenTaskTest.DummyGen.ALL_TARGETS = target_roots
+    return self._create_subclass_task(SimpleCodegenTaskTest.DummyGen,
+                                      target_roots=target_roots,
+                                      **options)
+
+  def test_codegen_strategy(self):
+    task = self.create_task(self.context())
+
+    self.set_options_for_scope(task.options_scope, strategy='global')
+    task = self.create_task(self.context())
+    self.assertEqual('global', task.get_options().strategy)
+    self.assertEqual('global', task.codegen_strategy.name)
+
+    self.set_options_for_scope(task.options_scope, strategy='isolated')
+    task = self.create_task(self.context())
+    self.assertEqual('isolated', task.codegen_strategy.name)
+
+    class TestGen(SimpleCodegenTask):
+      """Minimal class to test that forced codegen strategy works as intended."""
+
+      @property
+      def forced_codegen_strategy(self):
+        return 'global'
+
+    task = self._create_subclass_task(TestGen, strategy='global')
```

Comment:
Again, you could dispense with `_create_subclass_task()`  by just creating a second test class and have it return `TestGen` from its `task_type()` method.  My concern is that `_create_subclass_task()` re-invents some of the tricky stuff in the TaskTestBase superclass .